_posts/2022-12-16-aave-rebasing.md

```diff
@@ -0,0 +1,332 @@
+super.balanceOf(user).rayMul(
+  POOL.getReserveNormalizedIncome(_underlyingAsset)
+);
+{% endhighlight %}
+
+`rayMul` just means [multiply these values together then
+divide by a
+ray](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/math/WadRayMath.sol#L65-L74).
+
+So this `balanceOf` function is simply multiplying two values
+together. The first should be obvious: [it's just the user's balance in
+storage](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/base/IncentivizedERC20.sol#L106-L108).
+
+The second is less obvious. The logic for the getReserveNormalizedIncome function [looks like
+this](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/libraries/logic/ReserveLogic.sol#L40-L64) (again, with indentation):
+
+{% highlight javascript %}
+MathUtils.calculateLinearInterest(
+  reserve.currentLiquidityRate,
+  reserve.lastUpdateTimestamp
+).rayMul(
+  reserve.liquidityIndex
+);
+{% endhighlight %}
+
+So, the aToken `balanceOf` function is simply multiplying the stored
+balance by the current liquidity index (incremented by whatever interest has not
+yet been factored in).
+
+Initially, this cancels out the effect of scaling down the balance, so that if
+someone went to withdraw immediately after depositing they would get all of
+their money back. E.g. a stored balance of
+98.93e18 aPolyDAI would be multiplied by the liquidity index of 1.01083, resulting a
+withdrawable balance of 100 DAI.
+
+But over time, as the liquidity index increases, it results in a net positive aToken balance, since the stored balance is being multiplied by a larger number than the deposit was initially divided by.
+
+#### A Toy Example
+
+Let's put these concepts together and see how interest works in a bigger
+context.
+
+* suppose some asset has earned 5% interest over its lifetime in Aave
+* the liquidity index for this asset will be 5%, or 1.05e27
+* Ben deposits $100 of the asset
+* the aToken scales down his deposit by the liquidity index and stores the result,
+  in this case 100 / 1.05 == 95.23
+* if Ben immediately withdrew his money, his rebased balance would be: `storedBalance * liquidityIndex`, i.e. 95.23 * 1.05 == $100, what we expect
+* a year goes by and the pool earns another 5% interest, so the new liquidity index is 1.05 * 1.05 == 1.1025
+* I now deposit $100 of the asset
+* once again, the aToken scales down my deposit and stores the result: 100 / 1.1025 == 90.70 is my balance in storage
+* if I immediately withdrew, my rebased balance would be 90.70 * 1.1025 == $100, i.e. exactly what I put in
+* Ben's rebased balance at this point would be 95.23 * 1.1025 == $105
+* so he'd have 5% more underlying than me, as expected (since the pool as a whole earned 5% while he had his money in it)
+
+And that's it! That is how interest works in Aave.
+
+### Flexible Voting
```

Comment:
I think this high-level section should be first, then the technical details after?

Reply:
I think that's reasonable. Done in 8468901